### PR TITLE
Completed Farmar!

### PR DESCRIPTION
**Pull request by daphnegold**
https://farmar-rdarg.herokuapp.com/

Thank you for visiting FarMar!

Some shaved ice for Kari 🍧
Some flan for Charles 🍮

### CODE REVIEW

**Comment by kariabancroft** [comment]
Merci for the shaved ice! 🍧

File: app/controllers/markets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+class MarketsController < ApplicationController
+
+  before_action only:[:show, :edit] { @market = Market.find(params[:id]) }
```
Comment (kariabancroft) [inline app/controllers/markets_controller.rb:3]: Nice use of the before action! This is a lot in one line though, so I may extract the `find` logic out into a separate method that is just called in the `before_action`

File: app/controllers/markets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  def edit
+    @action = "update"
+    render :new
+  end
+
+  def update
+    updated_market = Market.update(params[:id], market_params)
+    redirect_to market_path(updated_market)
```
Comment (kariabancroft) [inline app/controllers/markets_controller.rb:26]: Nice use of the named routes in the controller

File: app/controllers/products_controller.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    vendor_id = params[:vendor_id]
+    redirect_to vendor_path(vendor_id)
+  end
+
+  ####################
+  private
+
+  def product_params
+    product = params.require(:product).permit(:name)
+    product.merge(params.permit(:vendor_id))
```
Comment (kariabancroft) [inline app/controllers/products_controller.rb:31]: Slick use of the `merge`!

File: app/controllers/vendors_controller.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
+class VendorsController < ApplicationController
+
+  before_action only:[:show, :edit] { @vendor = Vendor.find(params[:id]) }
+
+  def index
+    @market = Market.find(params[:market_id])
+  end
+
+  def show
+    if params[:product_select].present?
```
Comment (kariabancroft) [inline app/controllers/vendors_controller.rb:10]: I like the way you've used the `present?` method here

File: app/helpers/sales_helper.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+module SalesHelper
+  def in_dollars(sale)
```
Comment (kariabancroft) [inline app/helpers/sales_helper.rb:2]: Nice! Excellent extraction of this logic to somewhere other than the views/controllers

File: app/models/vendor.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    return total
+  end
+
+  def monthly_sales
+    month_sales = []
+
+    self.products.each do |product|
+      sales = product.sales.find_all do |sale|
+        unless sale.purchase_time.nil?
+          Time.now.month == sale.purchase_time.month && Time.now.year == sale.purchase_time.year
```
Comment (kariabancroft) [inline app/models/vendor.rb:19]: Seems like a lot of logic happening in this one line. I would recommend splitting out `Time.now` into a local variable and possibly purchase time as well so the month/year comparison is more clear.

File: app/views/markets/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
+<div class="center">
+  <%= form_for @market, url: { action: @action } do |f| %>
```
Comment (kariabancroft) [inline app/views/markets/new.html.erb:2]: Nice job making this dynamic for more than just the new action. I'd recommend moving this out into the `_form` partial and then calling this from within the `new` and `edit` views. It is being used for more than just the `new` action so it doesn't necessarily make sense to only be in the `new` view.

File: app/views/products/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<div class="center">
+  <%= form_for @product, url: { action: @action } do |f| %>
```
Comment (kariabancroft) [inline app/views/products/new.html.erb:2]: Same comment re: abstracting out from the `new` view.

File: app/views/sales/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<div class="center">
+  <%= form_for @sale, url: { action: @action } do |f| %>
```
Comment (kariabancroft) [inline app/views/sales/new.html.erb:2]: Same comment re: abstracting out from the `new` view.

File: app/views/vendors/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+  <h2>Sales</h2>
+  <ul>
+    <% @vendor.products.each do |product| %>
+    <li>
+      <span><%= link_to product.name, vendor_product_path(@vendor, product) %></span>
+    </li>
+    <% product.sales.each do |sale| %>
+      <div>
+        <%= number_to_currency(in_dollars sale.amount) %>
```
Comment (kariabancroft) [inline app/views/vendors/show.html.erb:27]: I would recommend using the `()` in your `in_dollars` method call to keep consistent with the way you call all other methods.

File: config/routes.rb [hunk 11]
```diff
@@ -1,4 +1,19 @@
 Rails.application.routes.draw do
+  root 'welcome#index'
```
Comment (kariabancroft) [inline config/routes.rb:2]: I'm glad you have the root route with the welcome controller set up

**Comment by kariabancroft** [comment]
@daphnegold & @rmtolmach - you did a really excellent job with the separation of concerns in this project. Your views and controllers are clean with the use of model & helper methods. Nice job!

**Comment by daphnegold** [comment]
Ty for the comments, Kari!

**Pull request closed without merging**